provider/openstack/firewaller.go

```diff
@@ -545,92 +545,139 @@ func (c *neutronFirewaller) setUpGlobalGroup(groupName string, apiPort int) (neu
+		rule.ParentGroupId = group.Id
 		// Neutron translates empty RemoteIPPrefix into
 		// 0.0.0.0/0 or ::/0 instead of ParentGroupId
 		// when EthernetType is set
 		if rule.RemoteIPPrefix == "" {
-			rule.RemoteGroupId = newGroup.Id
+			rule.RemoteGroupId = group.Id
 		}
-		groupRule, err := neutronClient.CreateSecurityGroupRuleV2(rule)
+		_, err := neutronClient.CreateSecurityGroupRuleV2(rule)
 		if err != nil {
 			return zeroGroup, err
 		}
-		newGroup.Rules = append(newGroup.Rules, *groupRule)
 	}
-	return *newGroup, nil
-}
 
-func countIngressRules(rules []neutron.SecurityGroupRuleV2) int {
-	count := 0
-	for _, rule := range rules {
-		if rule.Direction == "ingress" {
-			count += 1
-		}
+	// Since we may have done a few add or delete rules, get a new
+	// copy of the security group to return containing the end
+	// list of rules.
+	groupsFound, err = neutronClient.SecurityGroupByNameV2(name)
+	if err == nil && len(groupsFound) == 1 {
+		group = groupsFound[0]
 	}
-	return count
+	return group, nil
 }
 
-// verifyGroupRules verifies the group rules against the rules we're looking for.
-func (c *neutronFirewaller) verifyGroupRules(rules []neutron.SecurityGroupRuleV2, rulesToMatch []neutron.RuleInfoV2) bool {
-	if countIngressRules(rules) != len(rulesToMatch) {
-		return false
-	}
-	count := len(rulesToMatch)
-	for _, rule := range rules {
-		// This is one of the default rules created when a new
-		// Neutron Security Group is created
-		if rule.Direction == "egress" {
-			continue
+// ruleInfoSet represents a Security Group Rule created for a Security Group.
+type ruleId struct {
+	found  bool
+	ruleId string
+}
+type ruleInfoSet map[neutron.RuleInfoV2]ruleId
+
+// newRuleSetForGroup returns a set of all of the permissions in a given
+// slice of SecurityGroupRules.  It ignores the group id, the
+// remove group id, and tenant id.  Keep the rule id to delete the rule if
+// necessary.
+func newRuleInfoSetFromRules(rules []neutron.SecurityGroupRuleV2) ruleInfoSet {
+	m := make(ruleInfoSet)
+	for _, r := range rules {
+		k := neutron.RuleInfoV2{
+			Direction:      r.Direction,
+			EthernetType:   r.EthernetType,
+			RemoteIPPrefix: r.RemoteIPPrefix,
 		}
-		for _, toMatch := range rulesToMatch {
-			var maxInt int
-			if rule.PortRangeMax != nil {
-				maxInt = *rule.PortRangeMax
-			} else {
-				maxInt = 0
-			}
-			var minInt int
-			if rule.PortRangeMin != nil {
-				minInt = *rule.PortRangeMin
-			} else {
-				minInt = 0
-			}
-			if rule.Direction == toMatch.Direction &&
-				rule.RemoteIPPrefix == toMatch.RemoteIPPrefix &&
-				*rule.IPProtocol == toMatch.IPProtocol &&
-				minInt == toMatch.PortRangeMin &&
-				maxInt == toMatch.PortRangeMax {
-				count -= 1
-				break
-			}
+		if r.IPProtocol != nil {
+			k.IPProtocol = *r.IPProtocol
 		}
+		if r.PortRangeMax != nil {
+			k.PortRangeMax = *r.PortRangeMax
+		}
+		if r.PortRangeMin != nil {
+			k.PortRangeMin = *r.PortRangeMin
+		}
+		v := ruleId{
+			found:  true,
```

Comment:
I can't see where a ruleId is constructed that ever has found = false
If so, why do we need found?